unittest: add COMPILER_WARNINGS_AS_ERRORS handling

fabiobaltieri opened the pull request:
Add a check to set the warnings_as_errors flags in unit tests if CONFIG_COMPILER_WARNINGS_AS_ERRORS is specified, this should catch warnings in unit test twister runs.

fabiobaltieri committed:
unittest: add COMPILER_WARNINGS_AS_ERRORS handling

Add a check to set the warnings_as_errors flags in unit tests if
CONFIG_COMPILER_WARNINGS_AS_ERRORS is specified, this should catch
warnings in unit test twister runs.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
tests: fix various unit tests warnings

Fix various unit test build warnings, unused variables, set but unused
variables, wrong types etc...

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

fabiobaltieri committed:
linker: undefine the no-ops dt macros at the end of the file

Undefine the DT_NODE_HAS_STATUS and DT_NODE_HAS_STATUS_OKAY no nop at
the end of the file so other uses can include devicetree.h without
generating a compiler warning.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

Thalley commented on subsys/bluetooth/host/conn.c:
```diff
@@ -139,7 +139,7 @@ void frag_destroy(struct net_buf *frag)
 	bt_buf_destroy_view(frag, &get_frag_md(frag)->view_meta);
 
-	LOG_DBG("");
+	LOG_DBG("%s", __func__);
```
What was the warning here? Empty log_dbg strings are fairly common because it logs the name of the function by default. I see at least 500 cases of this in the tree

Thalley replied:
Huh. 

This is going to be a growing pain if we don't fix the root cause. We don't want to apply this fix to all 500+ cases in the codebase. 

I would be OK with approving this as is, given that it would otherwise block this PR, but we would need a better solution for `LOG_DBG("")` for the `unit_testing` board

fabiobaltieri replied:
hang on, gcc and clang sets `-Wno-format-zero-length`, let's set it for unit test as well

fabiobaltieri replied:
ok added a flag, not sure it's the right/best location but I guess the buildsystem folks will point that out

Thalley commented on tests/bluetooth/audio/cap_commander/src/test_broadcast_reception.c:
```diff
@@ -104,13 +104,12 @@ static void cap_commander_test_broadcast_reception_before(void *f)
 
 	/* We need to cleanup since the CAP commander remembers state */
-	err = bt_cap_commander_cancel();
+	bt_cap_commander_cancel();
```
```suggestion
	(void)bt_cap_commander_cancel();
```

fabiobaltieri replied:
oky, done

Thalley commented on tests/bluetooth/audio/codec/src/main.c:
```diff
@@ -1701,7 +1701,7 @@ ZTEST(audio_codec_test_suite, test_bt_audio_codec_cap_meta_set_lang)
 		BT_HCI_CODING_FORMAT_LC3, 0x0000, 0x0000, {},
 		{BT_AUDIO_CODEC_DATA(BT_AUDIO_METADATA_TYPE_LANG, 'e', 'n', 'g')});
-	char new_expected_data[] = "deu";
+	unsigned char new_expected_data[] = "deu";
```
Hmm, this is a bit tricky. Technically `bt_audio_codec_cap_meta_set_lang` takes a `uint8_t` array which is of course unsigned, which I assume was the issue. 

However, we assign a string to it, and `string.h` and most string-related functions use `char` (without unsigned) as the type. 

By making this `unsigned char` it would now conflict with any string functions, which isn't really a problem in this case, but would be in many other usecases. 

If we really want to be type-true here, the type should probably be `const uint8_t []` instead (ditto for the above)

fabiobaltieri replied:
whatever makes you and the compiler happy :-) changed to to `const uint8_t[]`

kartben reviewed (dismissed):
Very nice, thanks for this!

Thalley commented on tests/unit/pot/nhpot.cpp:
```diff
@@ -39,7 +39,7 @@ ZTEST(pot, test_constexpr_NHPOT)
 
 	zassert_equal(64, val);
-	zassert_equal(64, nhpot(42));
+	zassert_equal(64, nhpot(42ULL));
```
Should this be 
```suggestion
	zassert_equal(64ULL, nhpot(42ULL));
```

as well then if we want to be using "correct" types?

fabiobaltieri replied:
sounds good, done

fabiobaltieri committed:
unittest: set -Wno-format-zero-length

This flag is set on other compiler as well for normal builds, and
suppresses a warning for empty print, which is apparently quite common.

The unit test build uses its own flags so set it there as well to get
the same behavior as with the rest of the code base.

Signed-off-by: Fabio Baltieri <fabiobaltieri@google.com>

jhedberg commented on subsys/bluetooth/host/hci_core.h:
```diff
@@ -7,7 +7,10 @@
  */
 
+/* Skip devicetree.h for unit test as it causes some define collitions. */
+#ifndef ZTEST_UNITTEST
 #include <zephyr/devicetree.h>
+#endif
```
Wouldn't it make more sense to have this #ifndef inside the `devicetree.h` file itself, so that not everybody who wants to include it needs to know of this exception?

fabiobaltieri replied:
Yeah not liking that either, but hey seems like that override was only needed on the linker file itself, I added a couple of undef for those at the end of the file and that seems to work, so that crappy hack is self contained at least.
